### PR TITLE
feat(bmap-tools_git.bb): recipe for bmap tools

### PR DESCRIPTION
**Pull request by sotaoverride**
Add bmap-tools for flashing whats on sd-cards to emmc. This would allow setting up the partition layout to
support A/B partition rootfs upgrades.

### CODE REVIEW

**Comment by DerekMaggio** [comment]
> Add bmap-tools for flashing whats on sd-cards to emmc. This would allow setting up the partition layout to
> support A/B partition rootfs upgrades.

Jeez I have no idea what any of this means. I am going to try to research it and then paraphrase it and you tell me how wrong I am.

You are adding `bmap-tools` for copying stuff from the SD card to the EMMC. 
An EMMC is a storage device soldered directly onto the board. Akin to a disk on a PC.
You are doing this because we want to support A/B partition. A/B partition is when you upload an upgrade to a different partition on disk and then boot to that partition. I assume this is valuable because you can still be running the current software while downloading and installing the upgrade to the second partition.

**Comment by sotaoverride** [comment]
> > Add bmap-tools for flashing whats on sd-cards to emmc. This would allow setting up the partition layout to
> > support A/B partition rootfs upgrades.
> 
> Jeez I have no idea what any of this means. I am going to try to research it and then paraphrase it and you tell me how wrong I am.
think you got it all figured out.  bmap-tool just does a better job at copying - doesnt leave any empty spaces in-between, theres also some other things like signing etc. For me it was just the new standard which yocto supports, so I went for it
> 
> You are adding `bmap-tools` for copying stuff from the SD card to the EMMC.
yup
> An EMMC is a storage device soldered directly onto the board. Akin to a disk on a PC.
yeah, I just htink of it as a sd-card soldered to the baord
> You are doing this because we want to support A/B partition. A/B partition is when you upload an upgrade to a different partition on disk and then boot to that partition. I assume this is valuable because you can still be running the current software while downloading and installing the upgrade to the second partition.
More like just have the working (current) partition to fall back on if it turns out the new rootfs has issues
these partitions would just have the root filesystem (/bin /usr /etc and all)

**Comment by DerekMaggio** [comment]
Okay that makes sense to me then. 
And the `bb` file is something that is generated by a command that says, "Give me `bmap-tools` and all of it's related dependencies"?

**Comment by sotaoverride** [comment]
> Okay that makes sense to me then.
> And the `bb` file is something that is generated by a command that says, "Give me `bmap-tools` and all of it's related dependencies"?

yeah, these .bb are bitbake recipes, which basically just bring in packages and all their dependencies and throws them on the target rootfs.. I get the basic structure for these recipes going using this command 
```devtool add <url>

File: recipes-ot3-tools/bmap-tools/bmap-tools_git.bb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+#   debian/copyright
+# NOTE: Original package / source metadata indicates license is: GPLv2
+#
+# NOTE: multiple licenses have been detected; they have been separated with &
+# in the LICENSE value for now since it is a reasonable assumption that all
+# of the licenses apply. If instead there is a choice between the multiple
+# licenses then you should change the value to separate the licenses with |
+# instead of &. If there is any doubt, check the accompanying documentation
+# to determine which situation is applicable.
+LICENSE = "GPLv2 & Unknown"
```
Comment (sfoster1) [inline recipes-ot3-tools/bmap-tools/bmap-tools_git.bb:20]: Let's get these sorted from the source please. The repo identifies itself on github and in COPYING as gpl-2 only.

File: recipes-ot3-tools/bmap-tools/bmap-tools_git.bb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+S = "${WORKDIR}/git"
+
+# NOTE: spec file indicates the license may be "GPL-2.0"
+inherit setuptools3
+
+# WARNING: the following rdepends are determined through basic analysis of the
+# python sources, and might not be 100% accurate.
+RDEPENDS_${PN} += "python3-compression python3-core python3-crypt python3-datetime python3-fcntl python3-io python3-logging python3-math python3-misc python3-mmap python3-six python3-unittest python3-xml"
+
+# WARNING: We were unable to map the following python package/module
```
Comment (sfoster1) [inline recipes-ot3-tools/bmap-tools/bmap-tools_git.bb:39]: We can remove some of these; the `six` deps should be handled by `python3-six`. `unittest2` and `mock` are for tests obviously. Not sure about `gpgme` and `backports`.

**Review by sfoster1 — CHANGES_REQUESTED** [review]
I think we don't want this as a `_git` recipe, which as far as I know is for development where the target version is likely to frequently change. I think we want this as `bmap-tools_3.6.bb` while keeping the `SRC_URI` as a git link.

**Comment by sotaoverride** [comment]
> > Okay that makes sense to me then.
> > And the `bb` file is something that is generated by a command that says, "Give me `bmap-tools` and all of it's related dependencies"?
> 
> yeah, these .bb are bitbake recipes, which basically just bring in packages and all their dependencies and throws them on the target rootfs.. I get the basic structure for these recipes going using this command

just realized the command I mentioned in my last comment didnt really show up here.. it's:
devtool add <github / pypi etc url>

File: recipes-ot3-tools/bmap-tools/bmap-tools_git.bb (hunk 1 above)
Comment (sotaoverride) [reply recipes-ot3-tools/bmap-tools/bmap-tools_git.bb:20]: ok, i guess I can leave both the license files in, but not call the other one unkown

File: recipes-ot3-tools/bmap-tools/bmap-tools_git.bb (hunk 2 above)
Comment (sotaoverride) [reply recipes-ot3-tools/bmap-tools/bmap-tools_git.bb:39]: seems to be ok, maybe gpgma and backports gets picked up becuase of some older version python support this package might be offering .. unclear.

Comment (sotaoverride) [reply recipes-ot3-tools/bmap-tools/bmap-tools_git.bb:39]: meant to say those six* packages do get taken care of by python3-six, so taking those out..

File: recipes-ot3-tools/bmap-tools/bmap-tools_3.6.bb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# Recipe created by recipetool
+# This is the basis of a recipe and may need further editing in order to be fully functional.
+# (Feel free to remove these comments when editing.)
+
+SUMMARY = "Tools to generate block map (AKA bmap) and flash images using bmap"
+# WARNING: the following LICENSE and LIC_FILES_CHKSUM values are best guesses - it is
```
Comment (sfoster1) [inline recipes-ot3-tools/bmap-tools/bmap-tools_3.6.bb:6]: Can we remove the autogenerated comments here and above

**Review by sfoster1 — APPROVED** [review]
Recipe looks good - do you want to add it to `IMAGE_INSTALL` here or in another pr

**Comment by sotaoverride** [comment]
> Recipe looks good - do you want to add it to `IMAGE_INSTALL` here or in another pr

Adding two packages to image recipe in a follow up pr

**Pull request merged**